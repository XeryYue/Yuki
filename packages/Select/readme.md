## how to do

3. fix window resize the option is dislocation

- [x] 