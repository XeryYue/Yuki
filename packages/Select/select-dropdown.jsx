import { defineComponent, watchEffect, ref, Teleport, nextTick } from 'vue'

import { useRect, useProvider, useResize } from '../utils'

const READONLY_SELECT_KEY = 'selectKey'

const getRect = (ref) => {
  const rect = useRect(ref)
  const { width, height, top, left } = rect
  return {
    width,
    height,
    left,
    top: top + height,
  }
}

const SelectDropDown = defineComponent({
  setup(props, { slots }) {
    const dropDownRef = ref(null)
    const { ctx } = useProvider(READONLY_SELECT_KEY)
    const { width, height } = useResize()

    const setPosition = () => {
      nextTick(() => {
        const rect = getRect(ctx.selectRef)
        const { width, top, left } = rect

        const style = {
          width: `${width}px`,
          top: `${top + 2}px`,
          left: `${left}px`,
        }
        dropDownRef.value.style.width = style.width
        dropDownRef.value.style.top = style.top
        dropDownRef.value.style.left = style.left
      })
    }
    watchEffect(() => {
      if (width.value || height.value) {
        setPosition()
      }
    })

    return () => (
      <Teleport to="body">
        <div
          class="fect-select__dropdown"
          ref={dropDownRef}
          v-show={ctx.visible.value}
        >
          {slots.default?.()}
        </div>
      </Teleport>
    )
  },
})

export default SelectDropDown
