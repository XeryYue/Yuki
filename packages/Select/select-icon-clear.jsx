import { defineComponent } from 'vue'

const SelcetClearableIcon = defineComponent({
  setup() {
    return () => (
      <div class="fect-select__clearIcon">
        <svg
          viewBox="0 0 24 24"
          stroke="currentColor"
          strokeWidth="1.5"
          strokeLinecap="round"
          strokeLinejoin="round"
          fill="none"
          shapeRendering="geometricPrecision"
          style={style.value}
        >
          <path d="M18 6L6 18" />
          <path d="M6 6l12 12" />
        </svg>
      </div>
    )
  },
})

export default SelcetClearableIcon
