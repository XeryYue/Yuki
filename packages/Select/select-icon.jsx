import { defineComponent } from 'vue'

const SelectIcon = defineComponent({
  props: {
    className: String,
  },
  setup(props) {
    return () => (
      <div class={`fect-select__icon-context ${props.className}`}>
        <svg
          viewBox="0 0 24 24"
          width="20px"
          height="20px"
          strokeWidth="1"
          strokeLinecap="round"
          color="currentColor"
          stroke="currentColor"
          strokeLinejoin="round"
          fill="none"
          shapeRendering="geometricPrecision"
        >
          <path d="M6 9l6 6 6-6" />
        </svg>
      </div>
    )
  },
})

export default SelectIcon
