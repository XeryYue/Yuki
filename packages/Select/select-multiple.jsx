import { defineComponent } from 'vue'
import SelcetClearableIcon from './select-icon-clear'

const SelectMultiple = defineComponent({
  props: {
    clearable: Boolean,
  },
  emits: ['clear'],
  setup(props, { slots, emit }) {
    const clearHandler = (e) => {
      e.stopPropagation()
      e.preventDefault()
      emit('clear', e)
    }

    return () => (
      <div class="fect-select__item">
        {slots.default?.()}
        {props.clearable && <SelcetClearableIcon onClick={clearHandler} />}
      </div>
    )
  },
})

export default SelectMultiple
