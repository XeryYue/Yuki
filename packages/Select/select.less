.fect-select {
  display: inline-flex;
  align-items: center;
  user-select: none;
  white-space: nowrap;
  max-width: 80vw;
  cursor: pointer;
  width: initial;
  overflow: hidden;
}
