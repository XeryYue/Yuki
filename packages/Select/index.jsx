import { computed, ref, watchEffect, toRefs } from 'vue'
import {
  createNameSpace,
  createProvider,
  theme,
  validator,
  useEventListener,
} from '../utils'
import SelectMultiple from './select-multiple'
import SelectIcon from './select-icon'
import SelcetClearableIcon from './select-icon-clear'
import SelectDropDown from './select-dropdown'
import './select.less'

const [createComponent] = createNameSpace('Select')

const { normalSizes } = theme

const READONLY_SELECT_KEY = 'selectKey'

const querySelectSize = (size) => {
  const sizes = {
    mini: {
      height: 'calc(1 * var(--fay-gap))',
      fontSize: '12px',
      minWidth: '100px',
    },
    small: {
      height: 'calc(1.35 * var(--fay-gap))',
      fontSize: '12px',
      minWidth: '128px',
    },
    medium: {
      height: 'calc(1.688 * var(--fay-gap))',
      fontSize: '14px',
      minWidth: '160px',
    },
    large: {
      height: 'calc(1.688 * var(--fay-gap))',
      fontSize: '16px',
      minWidth: '200px',
    },
  }
  return sizes[size]
}

export default createComponent({
  props: {
    modelValue: {
      type: [String],
      default: '',
    },
    placeholder: {
      type: [String],
      default: '',
    },
    size: {
      type: String,
      default: 'medium',
      validator: validator.enums(normalSizes),
    },
    clearable: {
      type: Boolean,
      default: true,
    },
    disabled: Boolean,
    width: {
      type: String,
      default: 'initial',
    },
    multiple: Boolean,
  },
  emits: ['change', 'update:modelValue'],
  setup(props, { attrs, slots, emit }) {
    const selectRef = ref(null)
    const visible = ref(false)
    const isEmpty = ref(false)
    const showClear = ref(false)

    const { provider, children } = createProvider(READONLY_SELECT_KEY)

    const setVisbile = (state) => (visible.value = state)

    const setEmpty = (state) => (isEmpty.value = state)

    const setShowClear = (state) => (showClear.value = state)

    const initIsEmpty = () => {
      const hasValue = !!props.modelValue
      if (hasValue) return setEmpty(false)
      return setEmpty(true)
    }

    useEventListener(
      'mouseenter',
      (e) => {
        e.stopPropagation()
        e.preventDefault()
        setShowClear(true)
      },
      selectRef,
    )
    useEventListener(
      'mouseleave',
      (e) => {
        e.stopPropagation()
        e.preventDefault()
        setShowClear(false)
      },
      selectRef,
    )

    watchEffect(initIsEmpty)

    /**
     * use computed to set global css variable
     * In previous version . It use watchEffect to set style and
     * save in reactive . it's too trouble ,too stupid.
     */
    const setStyle = computed(() => {
      const { height, fontSize, minWidth } = querySelectSize(props.size)
      const style = {}
      style['--select-fontSize'] = fontSize
      style['--select-minWidth'] = minWidth
      style['--select-height'] = height
      style.width = props.width
      return style
    })

    const handleClick = (e) => {
      if (props.disabled) return
      e.stopPropagation()
      e.preventDefault()
      const show = visible.value
      setVisbile(!show)
    }

    const clearIconHandler = () => {
      setVisbile(false)
      updateModelValue('')
    }

    const updateModelValue = (value) => emit('update:modelValue', value)

    const setChange = (value) => emit('change', value)

    // useEventListener('click', () => setVisbile(false))

    /**
     * filter and collect checked label array
     */
    const queryChecked = computed(() =>
      children.filter((child) => child.value === props.modelValue),
    )

    /**
     *  Display control right icon in singleMode
     */

    const rightVisible = computed(() => {
      const { clearable, disabled, modelValue } = props
      const show = clearable && !disabled && modelValue && showClear.value
      return show
    })

    provider({
      ...toRefs(props),
      selectRef,
      visible,
      setVisbile,
      setChange,
      updateModelValue,
    })

    const renderPlaceHolder = () => {
      return (
        <span class="value fect-select__placeholder">{props.placeholder}</span>
      )
    }

    const renderSingleMode = () => {
      return (
        <span class="value" style="justify-content: space-between">
          {queryChecked.value.map((child) => (
            <SelectMultiple>{child.label}</SelectMultiple>
          ))}
        </span>
      )
    }

    const renderMultipleMode = () => {
      return <div>1</div>
    }

    return () => (
      <div
        class={`fect-select ${props.disabled ? 'disabled' : ''}`}
        ref={selectRef}
        style={setStyle.value}
        onClick={handleClick}
      >
        {isEmpty.value && renderPlaceHolder()}
        {!isEmpty.value && renderSingleMode()}
        <SelectDropDown v-slots={slots} />
        {!rightVisible.value && (
          <SelectIcon className={visible.value ? 'click' : ''} />
        )}
        {rightVisible.value && (
          <SelcetClearableIcon
            disabled={props.disabled}
            onClick={clearIconHandler}
          />
        )}
      </div>
    )
  },
})
