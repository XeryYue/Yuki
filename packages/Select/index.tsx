import { computed, toRefs, PropType } from 'vue'
import { createNameSpace } from '../utils'
import { NormalSizes } from '../utils/theme/propTypes'
import './select.less'

const [createComponent] = createNameSpace('Select')

export default createComponent({
  props: {
    modelValue: {
      type: [Array, String],
      default: () => [],
    },
    placeholder: {
      type: String,
      default: '',
    },
    multiple: Boolean,
    size: {
      type: String as PropType<NormalSizes>,
      default: 'medium',
    },
    width: {
      type: String,
      default: 'initial',
    },
    clearable: Boolean,
    disabled: Boolean,
  },
  emits: ['change'],
  setup(props, { attrs, slots, emit }) {
    return () => <div class="fect-select">
      
    </div>
  },
})
