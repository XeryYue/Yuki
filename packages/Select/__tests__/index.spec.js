import Select from '../index'
import Option from '../../SelectOpiton'
import { mount } from '@vue/test-utils'

describe('Select', () => {
  it('should be render as element', () => {
    const wrapper = mount(Select)
    expect(wrapper.html()).toMatchSnapshot()
    expect(() => wrapper.unmount()).not.toThrow()
  })

  it('should support different sizes', () => {
    const wrapper = mount({
      setup() {
        return { sizes: ['mini', 'small', 'medium', 'large'] }
      },
      render() {
        return (
          <>
            {this.sizes.map((size, idx) => (
              <Select size={size} key={idx}></Select>
            ))}
          </>
        )
      },
    })
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.findAll('.fect-select')[0].attributes('style')).toContain(
      '--select-fontSize: 12px;',
    )
    expect(wrapper.findAll('.fect-select')[1].attributes('style')).toContain(
      '--select-fontSize: 12px;',
    )
    expect(wrapper.findAll('.fect-select')[2].attributes('style')).toContain(
      '--select-fontSize: 14px;',
    )
    expect(wrapper.findAll('.fect-select')[3].attributes('style')).toContain(
      '--select-fontSize: 16px;',
    )
  })

  it('should be support hidden clearable icon', async () => {
    const wrapper = mount({
      setup() {
        return {
          test: '0',
        }
      },
      render() {
        return (
          <Select modelValue={this.test} clearable={false}>
            <Option label="Vue" value="0"></Option>
            <Option label="React" value="1"></Option>
          </Select>
        )
      },
    })
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.find('.fect-select__clearIcon').exists()).toBe(false)
    await wrapper.setProps({ clearable: true })
    expect(wrapper.find('.fect-select__clearIcon').exists()).toBe(true)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should support disabled select component', () => {
    const wrapper = mount(Select, { props: { disabled: true } })
    expect(wrapper.find('.disabled').exists()).toBe(true)
  })

  it('should be support custom placeholder tips', () => {
    const wrapper = mount(Select, {
      props: {
        placeholder: 'custom',
      },
    })
    expect(wrapper.find('.fect-select__placeholder').html()).toContain('custom')
  })
})
