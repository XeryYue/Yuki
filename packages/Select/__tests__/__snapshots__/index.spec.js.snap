// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select clearable icon should be work 1`] = `
<div class="fect-select " style="--select-fontSize: 14px; --select-minWidth: 160px; --select-height: calc(1.688 * var(--fay-gap));">
  <!----><span class="value" style="justify-content: space-between;"></span>
  <!--teleport start-->
  <!--teleport end-->
  <div class="fect-select__icon-context "><svg viewBox="0 0 24 24" width="20px" height="20px" strokeWidth="1" strokeLinecap="round" color="currentColor" stroke="currentColor" strokeLinejoin="round" fill="none" shapeRendering="geometricPrecision">
      <path d="M6 9l6 6 6-6"></path>
    </svg></div>
</div>
`;

exports[`Select should be render as element 1`] = `
<div class="fect-select " style="--select-fontSize: 14px; --select-minWidth: 160px; --select-height: calc(1.688 * var(--fay-gap));"><span class="value fect-select__placeholder"></span>
  <!---->
  <!--teleport start-->
  <!--teleport end-->
  <div class="fect-select__icon-context "><svg viewBox="0 0 24 24" width="20px" height="20px" strokeWidth="1" strokeLinecap="round" color="currentColor" stroke="currentColor" strokeLinejoin="round" fill="none" shapeRendering="geometricPrecision">
      <path d="M6 9l6 6 6-6"></path>
    </svg></div>
</div>
`;

exports[`Select should be support hidden clearable icon 1`] = `
<div class="fect-select " style="--select-fontSize: 14px; --select-minWidth: 160px; --select-height: calc(1.688 * var(--fay-gap));">
  <!----><span class="value" style="justify-content: space-between;"></span>
  <!--teleport start-->
  <!--teleport end-->
  <div class="fect-select__icon-context "><svg viewBox="0 0 24 24" width="20px" height="20px" strokeWidth="1" strokeLinecap="round" color="currentColor" stroke="currentColor" strokeLinejoin="round" fill="none" shapeRendering="geometricPrecision">
      <path d="M6 9l6 6 6-6"></path>
    </svg></div>
</div>
`;

exports[`Select should be support hidden clearable icon 2`] = `
<div class="fect-select " style="--select-fontSize: 14px; --select-minWidth: 160px; --select-height: calc(1.688 * var(--fay-gap));">
  <!----><span class="value" style="justify-content: space-between;"><!----><div class="fect-select__item">Vue</div><div class="fect-select__clearIcon" style="margin-right: 5px;"><svg viewBox="0 0 24 24" stroke="currentColor" strokeWidth="1.5" strokeLinecap="round" strokeLinejoin="round" fill="none" shapeRendering="geometricPrecision" color="currentColor"><path d="M18 6L6 18"></path><path d="M6 6l12 12"></path></svg></div><!----></span>
  <!--teleport start-->
  <!--teleport end-->
  <div class="fect-select__icon-context "><svg viewBox="0 0 24 24" width="20px" height="20px" strokeWidth="1" strokeLinecap="round" color="currentColor" stroke="currentColor" strokeLinejoin="round" fill="none" shapeRendering="geometricPrecision">
      <path d="M6 9l6 6 6-6"></path>
    </svg></div>
</div>
`;

exports[`Select should support different sizes 1`] = `
<div class="fect-select " style="--select-fontSize: 12px; --select-minWidth: 100px; --select-height: calc(1 * var(--fay-gap));"><span class="value fect-select__placeholder"></span>
  <!---->
  <!--teleport start-->
  <!--teleport end-->
  <div class="fect-select__icon-context "><svg viewBox="0 0 24 24" width="20px" height="20px" strokeWidth="1" strokeLinecap="round" color="currentColor" stroke="currentColor" strokeLinejoin="round" fill="none" shapeRendering="geometricPrecision">
      <path d="M6 9l6 6 6-6"></path>
    </svg></div>
</div>
<div class="fect-select " style="--select-fontSize: 12px; --select-minWidth: 128px; --select-height: calc(1.35 * var(--fay-gap));"><span class="value fect-select__placeholder"></span>
  <!---->
  <!--teleport start-->
  <!--teleport end-->
  <div class="fect-select__icon-context "><svg viewBox="0 0 24 24" width="20px" height="20px" strokeWidth="1" strokeLinecap="round" color="currentColor" stroke="currentColor" strokeLinejoin="round" fill="none" shapeRendering="geometricPrecision">
      <path d="M6 9l6 6 6-6"></path>
    </svg></div>
</div>
<div class="fect-select " style="--select-fontSize: 14px; --select-minWidth: 160px; --select-height: calc(1.688 * var(--fay-gap));"><span class="value fect-select__placeholder"></span>
  <!---->
  <!--teleport start-->
  <!--teleport end-->
  <div class="fect-select__icon-context "><svg viewBox="0 0 24 24" width="20px" height="20px" strokeWidth="1" strokeLinecap="round" color="currentColor" stroke="currentColor" strokeLinejoin="round" fill="none" shapeRendering="geometricPrecision">
      <path d="M6 9l6 6 6-6"></path>
    </svg></div>
</div>
<div class="fect-select " style="--select-fontSize: 16px; --select-minWidth: 200px; --select-height: calc(1.688 * var(--fay-gap));"><span class="value fect-select__placeholder"></span>
  <!---->
  <!--teleport start-->
  <!--teleport end-->
  <div class="fect-select__icon-context "><svg viewBox="0 0 24 24" width="20px" height="20px" strokeWidth="1" strokeLinecap="round" color="currentColor" stroke="currentColor" strokeLinejoin="round" fill="none" shapeRendering="geometricPrecision">
      <path d="M6 9l6 6 6-6"></path>
    </svg></div>
</div>
`;
